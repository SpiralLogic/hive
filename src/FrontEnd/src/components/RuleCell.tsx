import { FunctionComponent, h } from 'preact';
import Arrow, { Direction } from './Arrow';
import Cell from './Cell';
import Tile from './Tile';

export const RuleCell: FunctionComponent<{
  result?: 'correct' | 'incorrect';
  creature?: string;
  symbol?: string;
  zIndex?: number;
  class?: string;
  correct?: ([Direction, number | undefined] | Direction)[];
  incorrect?: ([Direction, number | undefined] | Direction)[];
}> = (props) => {
  const { correct, incorrect, symbol, zIndex, result, ...rest } = props;
  const getResultChar = () => {
    const char = symbol ? symbol : result === 'correct' ? 'âœ”' : 'x';
    switch (props.result) {
      case 'correct':
        return <span>{char}</span>;
      case 'incorrect':
        return <span>{char}</span>;
    }
    return '';
  };

  return (
    <Cell class={props.result} style={zIndex ? { zIndex } : undefined}>
      <Tile creature={props.creature} {...rest}>
        {getResultChar()}
      </Tile>
      {correct?.map((direction) => (
        <Arrow result="correct" direction={direction instanceof Array ? direction : [direction, undefined]} />
      ))}
      {incorrect?.map((direction) => (
        <Arrow
          result="incorrect"
          direction={direction instanceof Array ? direction : [direction, undefined]}
        />
      ))}
    </Cell>
  );
};
