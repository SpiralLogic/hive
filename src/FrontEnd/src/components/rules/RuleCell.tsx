import { FunctionComponent, h } from 'preact';
import Arrow, { Direction } from '../Arrow';
import Hexagon from '../Hexagon';
import Tile from '../Tile';

const RuleCell: FunctionComponent<{
  result?: 'correct' | 'incorrect';
  creature?: string;
  symbol?: string;
  zIndex?: number;
  class?: string;
  correct?: ([Direction, number | undefined] | Direction)[];
  incorrect?: ([Direction, number | undefined] | Direction)[];
}> = (props) => {
  const { correct, incorrect, symbol, zIndex, result, ...rest } = props;
  const getResultChar = () => {
    const char = symbol ? symbol : result === 'correct' ? <span>&#10003;</span> : <span>&#10008;</span>;
    switch (result) {
      case 'correct':
        return <span>{char}</span>;
      case 'incorrect':
        return <span>{char}</span>;
      default:
        return '';
    }
  };

  return (
    <Hexagon role="cell" class={result} style={zIndex ? { zIndex } : undefined}>
      <Tile {...rest}>{getResultChar()}</Tile>
      {correct?.map((direction) => (
        <Arrow result="correct" direction={direction instanceof Array ? direction : [direction, undefined]} />
      ))}
      {incorrect?.map((direction) => (
        <Arrow
          result="incorrect"
          direction={direction instanceof Array ? direction : [direction, undefined]}
        />
      ))}
    </Hexagon>
  );
};

RuleCell.displayName = 'RuleCell';
export default RuleCell;
