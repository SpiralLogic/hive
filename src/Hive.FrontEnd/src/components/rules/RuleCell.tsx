import { FunctionComponent } from 'preact';
import { JSXInternal } from 'preact/src/jsx';

import Arrow, { Direction } from '../Arrow';
import Hexagon from '../Hexagon';
import Tile from '../Tile';

type Result = 'correct' | 'incorrect';
type Arrows = ([Direction, number | undefined] | Direction)[];

const createArrows = (arrows: Arrows | undefined, style: Result) =>
  arrows?.map((a) => {
    const [direction, length] = Array.isArray(a) ? a : [a];
    return <Arrow key={direction} result={style} direction={direction} length={length} />;
  });

const getResultChar = (result?: Result, symbol?: string) => {
  if (symbol) return <span>{symbol}</span>;
  switch (result) {
    case 'correct':
      return <span>&#10003;</span>;

    case 'incorrect':
      return <span>&#10008;</span>;

    default:
      return null;
  }
};

const RuleCell: FunctionComponent<{
  result?: Result;
  creature?: string;
  symbol?: string;
  zIndex?: number;
  selected?: boolean;
  correctArrows?: Arrows;
  incorrectArrows?: Arrows;
  style?: JSXInternal.CSSProperties;
}> = (properties) => {
  const { correctArrows, incorrectArrows, symbol, zIndex, result, ...rest } = properties;
  if (properties.selected) {
    properties.zIndex = 4;
  }

  const classes = [];
  if (result) classes.push(result);
  if (!properties.creature && !result) classes.push('blank');

  return (
    <Hexagon
      role="cell"
      class={classes.length > 0 ? classes.join(' ') : undefined}
      style={zIndex ? { zIndex } : undefined}>
      <Tile {...rest}>{getResultChar(result, symbol)}</Tile>
      {createArrows(correctArrows, 'correct')}
      {createArrows(incorrectArrows, 'incorrect')}
    </Hexagon>
  );
};

RuleCell.displayName = 'RuleCell';
export default RuleCell;
