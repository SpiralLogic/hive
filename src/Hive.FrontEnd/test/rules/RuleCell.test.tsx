import { render } from '@testing-library/preact';
import RuleCell from '../../src/components/rules/RuleCell';

describe('<RuleCell>', () => {
  it('snapshot', () => {
    expect(render(<RuleCell />).baseElement).toMatchSnapshot();
  });
});
