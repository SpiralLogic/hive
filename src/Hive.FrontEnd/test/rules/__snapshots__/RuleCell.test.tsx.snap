// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule cell snapshot tests snapshot 1`] = `
<body>
  <div>
    <div
      class="blank"
      role="cell"
    >
      <svg
        viewBox="0 0 100 100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          href="#hex"
        />
      </svg>
      <div
        class="tile"
      >
        <svg
          viewBox="0 0 100 100"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            href="#hex"
          />
        </svg>
      </div>
    </div>
  </div>
</body>
`;
